### PR TITLE
Redo bladder cancer analysis

### PR DESCRIPTION
**Pull request by ajlee21**
This PR performs multiple analyses after resolving the issue found in https://github.com/greenelab/generic-expression-patterns/pull/50

The main changes in this PR:
1. Updating scripts to clean up `find_specific_genes_in_new_experiment.ipynb`
2. Create new results files comparing treatment-resistant vs no treatment for different cell lines (KU1919, RT112, all cell lines combined)
3. Add notebook ____ to plot volcano plots of the template (real) experiment and the simulated experiments to confirm that our simulation is generating similar structure (i.e. similar effect sizes and signficance)

Minor changes:
1. Update readme to describe how a user can run their own analysis

### CODE REVIEW

File: new_experiment/nbconverted/viz_expression_data.py [hunk 1]
```diff
@@ -0,0 +1,482 @@
...
+
+# Encode compendium
+normalized_compendium = normalized_compendium_data.drop(['sample group'], axis=1)
+compendium_encoded = loaded_model.predict_on_batch(normalized_compendium)
+
+compendium_encoded_df = pd.DataFrame(data=compendium_encoded, 
+                                     index=normalized_compendium.index)
+
+# Get and save PCA model
+model2 = pca.fit(compendium_encoded_df)
```
Comment (jjc2718) [inline new_experiment/nbconverted/viz_expression_data.py:387]: Any reason you're using PCA for the latent space and UMAP for the gene expression space?

**Review by jjc2718 — APPROVED** [review]
Nice work! I think the code looks fine so I'll approve, most of my inline comments are small things or conceptual questions.

These are definitely surprising results that I think may take a while to detangle. Here are some general questions/comments I have after looking at the slides you linked:

* When you say "template", you mean the data used to train the VAE, right? So each of the facet plots in your slide decks is a separate training dataset? Or is it possible to train the VAE on a large dataset (e.g. recount2) and seed the simulations with a smaller dataset somehow? I think you may have explained this in your past PRs but I can't remember exactly.
* It looks like it's fairly well known, [e.g. this paper,](https://arxiv.org/abs/2010.09042) that VAEs tend to underestimate data variance, and thus that they don't necessarily do a great job at modeling outliers/heavy tails (which I guess might be true of DEGs, especially in the case when there's only a few of them). Maybe you could look at the parameters your VAE is learning from the template data (specifically the variances or covariance matrix) - I'd expect the variance of the latent Gaussian distribution to be lower in the cases when there's only a few DEGs. This doesn't necessarily explain the extreme differences between simulated experiments (since IIRC all of the simulated experiments are generated from the same VAE, just with different linear shifts) but it might explain some of the difference between templates with many DEGs and few DEGs.
* I agree with the comments in one of the slide decks that it would be good to compare the shift vectors for the different simulated experiments, to see if there are any commonalities in the experiments that seem to collapse to a horizontal line in the volcano plot as opposed to the normal-looking ones.
* Do you have any idea what's different about the Powers et al. case? There's only 33 DEGs, but the simulated volcano plots look pretty normal/uniform (even though most of them don't have any differential expression).

File: generic_expression_patterns_modules/plot.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+        Number of columns in facet plot
+    nrows: int
+        Number of rows in facet plot
+    fig_width: int
+        Width of figure
+    fig_height: ing
+        Height of figure
+
+	"""
+    fig, axes = plt.subplots(ncols=ncols, nrows=nrows, figsize=(fig_width, fig_height))
```
[the remark this answers is not in the document]
Comment (ajlee21) [reply generic_expression_patterns_modules/plot.py:105]: Thank you for the tip. I did think about using a facet, but it looked like each facet was a variable of the data and in my case each facet is a different experiment so that is why I chose to do it with subplots. But I guess I could add a label for each experiment to wrap by that. Definitely something for the future..thanks!

File: generic_expression_patterns_modules/stats.py [hunk 3]
```diff
@@ -80,9 +80,9 @@ def format_pseudomonas_pathway_DB(pathway_DB_filename, local_dir, out_filename):
 
 def process_samples_for_limma(
     expression_filename,
-    process_metadata_filename,
     grp_metadata_filename,
     out_expression_filename=None,
+    process_metadata_filename=None,
```
[the remark this answers is not in the document]
Comment (ajlee21) [reply generic_expression_patterns_modules/stats.py:85]: Very good point. I will add this

File: new_experiment/nbconverted/viz_expression_data.py [hunk 4]
```diff
@@ -0,0 +1,482 @@
...
+    simulated_DE_stats_dir,
+    project_id,
+    pval_name,
+    logFC_name,
+    num_simulated,
+    5,
+    5,
+    20,
+    15
+)
```
[the remark this answers is not in the document]
Comment (ajlee21) [reply new_experiment/nbconverted/viz_expression_data.py:161]: My expectation would be that the simulated experiments and template experiment have similar **distribution of effect size changes** (i.e. our simulation preserve the experimental design, so the "changes" are maintained). We expect the shape of the volcano to be similar between the template and simulated

However, given what you mentioned earlier, about VAE reducing the variance in the data, I would expect a lower range of effect sizes and even a reduced number of DEGs in the simulated experiments.

File: new_experiment/nbconverted/viz_expression_data.py [hunk 5]
```diff
@@ -0,0 +1,482 @@
...
+
+print(simulated_data.shape)
+simulated_data
+
+
+# In[15]:
+
+
+sns.distplot(simulated_data.iloc[0:2].mean(), kde=False)
+sns.distplot(simulated_data.iloc[2:].mean(), kde=False)
```
[the remark this answers is not in the document]
Comment (ajlee21) [reply new_experiment/nbconverted/viz_expression_data.py:236]: I would have expected some reduction in the fold change, but not to such an extent as you mention. One possible explanation is that the transformation moves the experiment outside the bounds of what is well represented then this change can be decoded to be nearly 0, however the simulated experiments so far don't seem to be in a valley.

File: new_experiment/nbconverted/viz_expression_data.py [hunk 6]
```diff
@@ -0,0 +1,482 @@
...
+    legend_title=element_text(family='sans-serif', size=15),
+    legend_text=element_text(family='sans-serif', size=12),
+    plot_title=element_text(family='sans-serif', size=15),
+    axis_text=element_text(family='sans-serif', size=12),
+    axis_title=element_text(family='sans-serif', size=15)
+    )
+fig += scale_color_manual(['#bdbdbd', 'red', 'blue'])
+fig += guides(colour=guide_legend(override_aes={'alpha': 1}))
+
+print(fig)
```
[the remark this answers is not in the document]
Comment (ajlee21) [reply new_experiment/nbconverted/viz_expression_data.py:343]: The plots serve two purposes: 1) I wanted to confirm that the latent space shifting looked as expected (i.e. the structure of the experiments were preserved). This was to rule out a mistake in the simulation that might be the reason for the volcano plots looking not as expected. Since the shift is linear that is the reason I was using PCA. But I have also added UMAP to look at the latent space as a whole in addition to the first two PCs. 2) I wanted to confirm that the simulated experiments were within the bounds of the data as opposed to a valley that might explain the consistency in patterns.

File: new_experiment/nbconverted/viz_expression_data.py (hunk 1 above)
Comment (ajlee21) [reply new_experiment/nbconverted/viz_expression_data.py:387]: See comment above: PCA was used to test if the simulation was implemented propoerly -i.e. the simulation produced a linear shift of the template experiment

**Comment by ajlee21** [comment]
> When you say "template", you mean the data used to train the VAE, right? So each of the facet plots in your slide decks is a separate training dataset? Or is it possible to train the VAE on a large dataset (e.g. recount2) and seed the simulations with a smaller dataset somehow? I think you may have explained this in your past PRs but I can't remember exactly.

Sorry for the confusion. I should add more of a description about this. So the VAE is trained on a large dataset like recount2. The template experiment in this case is an individual experiment that is either one of the experiments in the training dataset or it can be a completely new experiment that is not included in the training set. This template experiment represents a real experiment. And the simulated experiments are generated by starting with this real template experiment, embedding it into the learned latent space and shifting it in that latent space. Each facet of the plot is the volcano plot for each of the 25 simulated experiments.

>It looks like it's fairly well known, e.g. this paper, that VAEs tend to underestimate data variance, and thus that they don't necessarily do a great job at modeling outliers/heavy tails (which I guess might be true of DEGs, especially in the case when there's only a few of them). Maybe you could look at the parameters your VAE is learning from the template data (specifically the variances or covariance matrix) - I'd expect the variance of the latent Gaussian distribution to be lower in the cases when there's only a few DEGs. This doesn't necessarily explain the extreme differences between simulated experiments (since IIRC all of the simulated experiments are generated from the same VAE, just with different linear shifts) but it might explain some of the difference between templates with many DEGs and few DEGs.

Thank you for the citations. This reduction in variance is something I am aware of but didn't have a publication to refer to so this is helpful. Your logic is definitely what Casey and I are both thinking might explain the inconsistency between the template vs simulated experiments for those with fewer DEGs. 

My only guess for the consistency amongst the simulated experiments, might depend on the location in the latent space the experiment was shifted since the latent space captures a wide range of experiments.

>I agree with the comments in one of the slide decks that it would be good to compare the shift vectors for the different simulated experiments, to see if there are any commonalities in the experiments that seem to collapse to a horizontal line in the volcano plot as opposed to the normal-looking ones.

Definitely need to figure out if it makes sense to do more testing or if there is a way for me to explain the observations as is since we are trying to wrap up this manuscript

>Do you have any idea what's different about the Powers et al. case? There's only 33 DEGs, but the simulated volcano plots look pretty normal/uniform (even though most of them don't have any differential expression).

The Powers et. al. dataset is more homogeneous compared to the Pseudomonas or recount2 datasets. I suspect this is possibly the reasons for the higher uniformity.

**Comment by jjc2718** [comment]
>  So the VAE is trained on a large dataset like recount2. The template experiment in this case is an individual experiment that is either one of the experiments in the training dataset or it can be a completely new experiment that is not included in the training set. This template experiment represents a real experiment. And the simulated experiments are generated by starting with this real template experiment, embedding it into the learned latent space and shifting it in that latent space. Each facet of the plot is the volcano plot for each of the 25 simulated experiments.

Ahh, got it. You can disregard what I said about the VAE latent distribution parameters, then - sounds like they'll all be the same since you're only training 1 VAE per species/compendium.

> My only guess for the consistency amongst the simulated experiments, might depend on the location in the latent space the experiment was shifted since the latent space captures a wide range of experiments.

It has to be a difference in the shift, doesn't it? Each simulated experiment starts with the same dataset but has a unique  shift vector, so there's nothing else that varies between simulated experiments, correct? Or is there another random component?

File: new_experiment/nbconverted/viz_expression_data.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jjc2718) [reply new_experiment/nbconverted/viz_expression_data.py:343]: Got it! That makes sense.

**Comment by ajlee21** [comment]
* The template experiment is shifted to a new location based on a random vector that is determined based on the distribution of the training compendium. 
* The Powers et. al. distribution looks very uniform, so any shifting would move the experiment to any equally populated location. And since all the experiments are of a similar design, any new location will have a very similar pattern.
* By comparison, shifting experiments in recount2 or Pseudomonas could result in the experiment being located in a less population region or a region representing different expression patterns. So the decoding of these shifted experiments will try to mostly preserve the structure of the template but may result in slight differences.
* To test this I would need to dive deeper into how the latent space is structured. Which is definitely very interesting but something that maybe should wait.

File: new_experiment/nbconverted/viz_expression_data.py [hunk 7]
```diff
@@ -0,0 +1,482 @@
...
+    logFC_name
+)
+
+
+# In[7]:
+
+
+simulated_DE_stats_dir = os.path.join(local_dir, "DE_stats")
+
+plot.make_volcano_plot_simulated(
```
[the remark this answers is not in the document]
Comment (ajlee21) [reply new_experiment/nbconverted/viz_expression_data.py:151]: Good catch. I will fix this

File: new_experiment/nbconverted/viz_expression_data.py [hunk 8]
```diff
@@ -0,0 +1,482 @@
...
+
+import os
+import pandas as pd
+import umap
+import pickle
+import glob
+import seaborn as sns
+from sklearn.decomposition import PCA
+from keras.models import load_model
+from plotnine import (ggplot,
```
[the remark this answers is not in the document]
Comment (ajlee21) [reply new_experiment/nbconverted/viz_expression_data.py:29]: I agree thanks for the tip to help me clean up this cell.
